Balancer-Bots/Large-Internal-Balance-Deposit-Withdrawal/src/balance.fetcher.ts

```diff
@@ -0,0 +1,39 @@
+import { providers, Contract, BigNumber } from "ethers";
+import { Interface } from "ethers/lib/utils";
+import LRU from "lru-cache";
+import { TOKEN_ABI } from "./constants";
+
+export default class BalanceFetcher {
+  provider: providers.Provider;
+  private cache: LRU<string, BigNumber>;
+  tokenAddress: string;
+  private tokenContract: Contract;
+
+  constructor(provider: providers.Provider) {
+    this.provider = provider;
+    this.cache = new LRU<string, BigNumber>({
+      max: 10000,
+    });
+    this.tokenAddress = "";
+    this.tokenContract = new Contract(this.tokenAddress, new Interface(TOKEN_ABI), this.provider);
+  }
+
+  // Main function to fetch the contract balance.
+  public async getBalance(block: number | string, vaultAddress: string, _tokenAddress: string): Promise<BigNumber> {
+    this.tokenAddress = _tokenAddress;
+    if (this.tokenContract.address != this.tokenAddress) {
+      this.tokenContract = this.tokenContract.attach(this.tokenAddress);
+    }
+
+    const key: string = `${this.tokenAddress}-${block}`;
+    if (this.cache.has(key)) return this.cache.get(key) as BigNumber;
+
+    const balance: BigNumber = await this.tokenContract.balanceOf(vaultAddress, {
+      blockTag: block,
+    });
+
+    this.cache.set(key, balance);
+
+    return balance;
+  }
+}
```

Comment:
```suggestion
  provider: providers.Provider;
  private cache: LRU<string, BigNumber>;
  private tokenContract: Contract;

  constructor(provider: providers.Provider) {
    this.provider = provider;
    this.cache = new LRU<string, BigNumber>({
      max: 10000,
    });
    this.tokenContract = new Contract("", new Interface(TOKEN_ABI), this.provider);
  }

  // Main function to fetch the contract balance.
  public async getBalance(block: number | string, vaultAddress: string, tokenAddress: string): Promise<BigNumber> {
    const token = this.tokenContract.attach(tokenAddress);

    const key: string = `${tokenAddress}-${block}`;
    if (this.cache.has(key)) return this.cache.get(key) as BigNumber;

    const balance: BigNumber = await token.balanceOf(vaultAddress, {
      blockTag: block,
    });

    this.cache.set(key, balance);

    return balance;
  }
}
```

In this case, I think just this is fine, no need to store the address.

Reply:
Besides that, I think it's fine already, so I'm approving.

Reply:
don't we still need to check if the current contract's address is the same with the token address in order to prevent new contract creation for the same token?

Reply:
Ohh, I get what you're talking about, in this way, you don't really have the address in the contract info, since `attach` creates a new contract with `this.tokenContract` info except for the address, which is specified. What I said about the cache still holds, but imo it's not worth it in this case.